### PR TITLE
Add an overload to UseCosmos that takes a connection string

### LINKED ISSUE #20712: Add support for Cosmos DB Connection strings
**Current State**
Currently, the `DbContextOptionsBuilder.UseCosmos` method has one overload that takes the Cosmos DB URI and key. 

**Potential Issue**
We're about to ship some updates in the tooling that will dial in the Cosmos DB connection string more elegantly. Customers who use that tooling would get a connection string handed to them in config, but they'd have a paper cut here - they can't use that pre-wired-up connection string with the `UseCosmos` method. 

**Proposed Solution**
Add an overload that takes a single string rather than the URI and the key, and take in the Connection String the tooling team will be pulling down from the Cosmos DB RP. 

Adding @timheuer and @AngelosP and @vijayrkn FYI for visibility. If you'd like support on the feature let us know.

### PR DESCRIPTION
**Pull request by Marusyk**
Add an overload to UseCosmos that takes a connection string.

Fixes #20712

For now, I have some problems with testing it. Could you please suggest if I'm on the right way? I'll continue to work on tests.

### CODE REVIEW

File: src/EFCore.Cosmos/Infrastructure/Internal/CosmosDbOptionExtension.cs [hunk 1]
```diff
@@ -134,6 +136,29 @@ public virtual CosmosOptionsExtension WithDatabaseName([NotNull] string database
...
+        ///     doing so can result in application failures when updating to a new Entity Framework Core release.
+        /// </summary>
+        public virtual CosmosOptionsExtension WithConnectionString([NotNull] string connectionString)
+        {
+            var clone = Clone();
+
+            clone._connectionString = connectionString;
+
+            return clone;
+        }
```
Comment (AndriySvyryd) [inline src/EFCore.Cosmos/Infrastructure/Internal/CosmosDbOptionExtension.cs:160]: Also update `GetServiceProviderHashCode` below

File: src/EFCore.Cosmos/Infrastructure/Internal/CosmosSingletonOptions.cs [hunk 2]
```diff
@@ -85,7 +94,8 @@ public virtual void Validate(IDbContextOptions options)
             var cosmosOptions = options.FindExtension<CosmosOptionsExtension>();
 
             if (cosmosOptions != null
-                && (AccountEndpoint != cosmosOptions.AccountEndpoint
+                && ConnectionString != cosmosOptions.ConnectionString
+                || (AccountEndpoint != cosmosOptions.AccountEndpoint
```
Comment (AndriySvyryd) [inline src/EFCore.Cosmos/Infrastructure/Internal/CosmosSingletonOptions.cs:98]: Move the parenthesis one line up.

File: src/EFCore.Cosmos/Storage/Internal/SingletonCosmosClientWrapper.cs [hunk 3]
```diff
@@ -61,7 +63,9 @@ public SingletonCosmosClientWrapper([NotNull] ICosmosSingletonOptions options)
         ///     any release. You should only use it directly in your code with extreme caution and knowing that
         ///     doing so can result in application failures when updating to a new Entity Framework Core release.
         /// </summary>
-        public virtual CosmosClient Client => _client ??= new CosmosClient(_endpoint, _key, _options);
+        public virtual CosmosClient Client => string.IsNullOrEmpty(_connectionString)
+                    ? (_client ??= new CosmosClient(_endpoint, _key, _options))
+                    : (_client ??= new CosmosClient(_connectionString, _options));
```
Comment (AndriySvyryd) [inline src/EFCore.Cosmos/Storage/Internal/SingletonCosmosClientWrapper.cs:68]: Move `_client ??=` outside the conditional

**Comment by AndriySvyryd** [comment]
General approach looks good. You can put the tests in a new class similar to https://github.com/dotnet/efcore/blob/bde2b140d6f4cf94d6d1285d402941e20193ec60/test/EFCore.SqlServer.FunctionalTests/ConnectionSpecificationTest.cs

**Comment by Marusyk** [comment]
> Add an overload that takes a single string rather than the URI and the key
It seems we cannot use a single string parameter because we also need to specify the database name.

The connection string to CosmosDB doesn't contains Database name

> AccountEndpoint=https://schindler-eos.documents.azure.com:443/;AccountKey=SIUFFZa48DZsbCSghWcULcU2vZHr8d8mtCBFGf9MtWd9aL71eYg4KUbSoefachZnJQQUcFtIBmXi345F4tsbDw==;

@AndriySvyryd

**Comment by AndriySvyryd** [comment]
> should I add something like:
UseCosmos(string connectionString, string database)

Yes, all overloads should have it

**Comment by Marusyk** [comment]
Added only one test for now. 
Not sure about changes in `GetServiceProviderHashCode` and `PopulateDebugInfo` methods from `CosmosOptionsExtension` file.

**Comment by Marusyk** [comment]
Could you please help with the build? Don't understand what's wrong, I got "Build succeeded." on the local machine.

**Comment by AndriySvyryd** [comment]
@Marusyk Rebase your changes on latest master

File: src/EFCore.Cosmos/Infrastructure/Internal/CosmosDbOptionExtension.cs [hunk 4]
```diff
@@ -253,12 +278,21 @@ public override long GetServiceProviderHashCode()
...
+                        hashCode = (hashCode * 397) ^ (Extension._connectionMode?.GetHashCode() ?? 0);
+                        _serviceProviderHash = hashCode;
+                    }
+                    else
+                    {
+                        var hashCode = Extension._accountEndpoint.GetHashCode();
+                        hashCode = (hashCode * 397) ^ Extension._accountKey.GetHashCode();
+                        hashCode = (hashCode * 397) ^ (Extension._region?.GetHashCode() ?? 0);
+                        hashCode = (hashCode * 397) ^ (Extension._connectionMode?.GetHashCode() ?? 0);
+                        _serviceProviderHash = hashCode;
```
Comment (AndriySvyryd) [inline src/EFCore.Cosmos/Infrastructure/Internal/CosmosDbOptionExtension.cs:294]: The last 3 lines are the same, so you can move them out of the block

File: test/EFCore.Cosmos.FunctionalTests/config.json [hunk 5]
```diff
@@ -2,7 +2,8 @@
   "Test": {
     "Cosmos": {
       "DefaultConnection": "https://localhost:8081",
-      "AuthToken": "C2y6yDjf5/R+ob0N8A7Cgv30VRDJIWEHLM+4QDU5DE2nQ9nDuVTqobD4b8mGGyPMbIZnqyMsEcaGQy67XIw/Jw=="
+      "AuthToken": "C2y6yDjf5/R+ob0N8A7Cgv30VRDJIWEHLM+4QDU5DE2nQ9nDuVTqobD4b8mGGyPMbIZnqyMsEcaGQy67XIw/Jw==",
+      "ConnectionString": "AccountEndpoint=https://localhost:8081/;AccountKey=C2y6yDjf5/R+ob0N8A7Cgv30VRDJIWEHLM+4QDU5DE2nQ9nDuVTqobD4b8mGGyPMbIZnqyMsEcaGQy67XIw/Jw=="
```
Comment (AndriySvyryd) [inline test/EFCore.Cosmos.FunctionalTests/config.json:6]: No need to specify this here, the default in `TestEnvironment` is enough

**Comment by AndriySvyryd** [comment]
@Marusyk Thanks for your contribution!

**Pull request merged**